src/diff/index.js

```diff
@@ -253,7 +258,7 @@ export function diff(
 			c.base = newVNode._dom;
 
 			// We successfully rendered this VNode, unset any stored hydration/bailout state:
-			newVNode._hydrating = null;
+			newVNode._flags &= RESET_MODE;
```

Comment:
This can be removed, I removed it in my PR and everything worked correctly

Reply:
Oh interesting. Would've expected this to error. Might suggest we are missing a test case. Gonna look into this to see understand that better.

Reply:
I think imma keep this cuz there is a situation where we'll preserve the MODE_HYDRATE, etc. flags causing rerenders of the component to go down the hydration code paths. Given how hydration works today, it doesn't appear to lead to incorrect DOM, just unnecessarily running through the hydration code paths.

Specifically the test I added in my last commit reproduces what I'm thinking. Normally with Suspense, when the Suspense rerenders on promise resolution, we'll generate a new VNode using createElement for the component that suspended, which will have `0` for `_flags`. Once that rerender completes successfully, the `0` for `_flags` will become the new tree, effectively "clearing" the flags on the suspending component/VNode.

However, if that suspending component is memoized, then when Suspense rerenders on promise resolution, we won't generate a new VNode for the suspending component using createElement. Instead will trigger a forceUpdate on the suspending component, which will generate VNode using `Object.assign` in `renderComponent`. The newVNode for the suspending component will have the same flags as the oldVNode, i.e. MODE_HYDRATE. Once it successfully rerenders, the newVNode will become part of the tree, preserving the MODE_HYDRATE in the new tree. As long as the component doesn't get any rerenders from the parent (which is likely since this component is memoized), it will only rerender by going through `renderComponent` invoking the same behavior as described here (i.e. creating a newVNode by copying oldVNode therefore inheriting flags).

However, since with MODE_HYDRATION set, we just set oldDom to newVNode._dom, it usually just works. We also set excessDomChildren to `[oldDom]` which is what invokes some hydration code paths. Presumably there is a situation where setting `oldDom` to `newVNode._dom` would be incorrect, but I haven't come up with one yet. But to be safer, Imma leave this line here.